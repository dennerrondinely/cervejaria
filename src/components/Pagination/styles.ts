import styled, { css } from 'styled-components';

type ButtonProps = {
  selected?: boolean;
};

export const Wrapper = styled.div`
  display: flex;
  align-items: center;
  width: 100%;
  margin: 0 auto;
  max-width: 30rem;
  justify-content: space-between;
`;

export const PagesWrapper = styled.div`
  display: flex;
  align-items: center;
  gap: 1rem;
`;

export const Page = styled.button`
  border: none;
  background-color: transparent;
  cursor: pointer;
  font-family: 'Roboto';
  font-style: normal;
  font-weight: 700;
  font-size: 18px;
  line-height: 21px;
  letter-spacing: 0.36px;

  color: #000000;

  ${({ selected }: ButtonProps) => css`
    font-weight: ${selected ? 700 : 400};
  `}
`;

export const Navigation = styled.button`
  border: none;
  background-color: transparent;
  cursor: pointer;

  &:disabled {
    opacity: 0.3;
  }
`;
