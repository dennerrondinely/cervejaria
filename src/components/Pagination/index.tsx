import Icon from 'components/Icon';
import { useEffect, useState } from 'react';
import getPagination from 'utils/getPagination';

import * as S from './styles';

type PaginationProps = {
  pageSize: number;
  total: number;
  onPaginate?(page: number): void;
};

const Pagination = ({ pageSize, total, onPaginate }: PaginationProps) => {
  const [pagination, setPaginatio] = useState<number[]>([]);
  const [select, setSelect] = useState(1);
  const pageNumbers = Math.ceil(total / pageSize);

  useEffect(() => {
    const pages = getPagination(2, pageNumbers);
    setPaginatio(pages);
  }, [pageNumbers]);

  const handlePagination = (number) => {
    const pages = getPagination(number, pageNumbers);
    setPaginatio(pages);
    setSelect(number);
    onPaginate && onPaginate(number);
  };

  const handlePrevious = () => {
    const previous = select - 1;
    if (previous >= 1) handlePagination(previous);
  };

  const handleNext = () => {
    const next = select + 1;
    if (next <= pagination[pagination.length - 1]) handlePagination(next);
  };

  return (
    <>
      {pageNumbers > 1 && (
        <S.Wrapper>
          <S.Navigation
            onClick={handlePrevious}
            disabled={select <= 1}
            title="previous"
          >
            <Icon icon="chevron-left" size={23} />
          </S.Navigation>
          <S.PagesWrapper>
            {pagination.map(
              (page, i) =>
                page && (
                  <S.Page
                    key={i}
                    type="button"
                    selected={page === select}
                    onClick={() =>
                      page.toString() !== '...' && handlePagination(page)
                    }
                  >
                    {page}
                  </S.Page>
                )
            )}
          </S.PagesWrapper>
          <S.Navigation
            disabled={select >= pagination[pagination.length - 1]}
            onClick={handleNext}
            title="next"
          >
            <Icon icon="chevron-right" size={23} />
          </S.Navigation>
        </S.Wrapper>
      )}
    </>
  );
};

export default Pagination;
