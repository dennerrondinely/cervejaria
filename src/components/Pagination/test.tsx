import Pagination from './index';
import { fireEvent, render, screen, waitFor } from 'utils/tests/helpers';

describe('<Pagination />', () => {
  it('should render the Pagination', () => {
    const { container } = render(<Pagination pageSize={25} total={50} />);

    expect(container).toContainHTML('1');
    expect(container).toContainHTML('2');
    expect(container).toContainHTML('button');

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should return value when click in page', async () => {
    const onChange = jest.fn();

    render(<Pagination pageSize={25} total={50} onPaginate={onChange} />);

    const page = screen.getByText('2');
    fireEvent.click(page);

    await waitFor(() => {
      expect(onChange).toBeCalledWith(2);
    });
  });

  it('should return value when click in next', async () => {
    const onChange = jest.fn();

    render(<Pagination pageSize={25} total={50} onPaginate={onChange} />);

    const next = screen.getByTitle('next');
    fireEvent.click(next);

    await waitFor(() => {
      expect(onChange).toBeCalledWith(2);
    });
  });

  it('should return value when click in previous', async () => {
    const onChange = jest.fn();

    render(<Pagination pageSize={25} total={50} onPaginate={onChange} />);

    const previous = screen.getByTitle('previous');
    const next = screen.getByTitle('next');

    fireEvent.click(next);
    fireEvent.click(previous);

    await waitFor(() => {
      expect(onChange).toBeCalledWith(1);
    });
  });
});
