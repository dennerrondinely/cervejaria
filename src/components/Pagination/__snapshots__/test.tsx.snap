// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> should render the Pagination 1`] = `
.c2 {
  height: 23px;
  width: 23px;
  height: 23px;
  width: 23px;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  margin: 0 auto;
  max-width: 30rem;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 1rem;
}

.c4 {
  border: none;
  background-color: transparent;
  cursor: pointer;
  font-family: 'Roboto';
  font-style: normal;
  font-weight: 700;
  font-size: 18px;
  line-height: 21px;
  -webkit-letter-spacing: 0.36px;
  -moz-letter-spacing: 0.36px;
  -ms-letter-spacing: 0.36px;
  letter-spacing: 0.36px;
  color: #000000;
  font-weight: 700;
}

.c5 {
  border: none;
  background-color: transparent;
  cursor: pointer;
  font-family: 'Roboto';
  font-style: normal;
  font-weight: 700;
  font-size: 18px;
  line-height: 21px;
  -webkit-letter-spacing: 0.36px;
  -moz-letter-spacing: 0.36px;
  -ms-letter-spacing: 0.36px;
  letter-spacing: 0.36px;
  color: #000000;
  font-weight: 400;
}

.c1 {
  border: none;
  background-color: transparent;
  cursor: pointer;
}

.c1:disabled {
  opacity: 0.3;
}

<div
  class="c0"
>
  <button
    class="c1"
    disabled=""
    title="previous"
  >
    <svg
      class="c2"
      size="23"
      viewBox="0 0 15 25"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M14.838 21.6311L5.75481 12.603L14.838 3.57486L12.0416 0.801514L0.142303 12.603L12.0416 24.4045L14.838 21.6311Z"
        stroke-width="0"
      />
    </svg>
  </button>
  <div
    class="c3"
  >
    <button
      class="c4"
      type="button"
    >
      1
    </button>
    <button
      class="c5"
      type="button"
    >
      2
    </button>
  </div>
  <button
    class="c1"
    title="next"
  >
    <svg
      class="c2"
      size="23"
      viewBox="0 0 16 25"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0.749535 21.6311L9.83268 12.603L0.749535 3.57486L3.54587 0.801514L15.4452 12.603L3.54587 24.4045L0.749535 21.6311Z"
        stroke-width="0"
      />
    </svg>
  </button>
</div>
`;
